File: MATTHEW/leetcode/AIcoding/3.DeepSeek-v3/cpp_solutions/1745-palindrome-partitioning-iv.cpp
class Solution {
public:
    bool checkPartitioning(string s) {
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false));

        for (int i = 0; i < n; ++i) {
            dp[i][i] = true;
        }

        for (int i = 0; i < n - 1; ++i) {
            if (s[i] == s[i + 1]) {
                dp[i][i + 1] = true;
            }
        }

        for (int len = 3; len <= n; ++len) {
            for (int i = 0; i + len - 1 < n; ++i) {
                int j = i + len - 1;
                if (s[i] == s[j] && dp[i + 1][j - 1]) {
                    dp[i][j] = true;
                }
            }
        }

        for (int i = 1; i < n - 1; ++i) {
            for (int j = i; j < n - 1; ++j) {
                if (dp[0][i - 1] && dp[i][j] && dp[j + 1][n - 1]) {
                    return true;
                }
            }
        }

        return false;
    }
};

File: leetcode/AIcoding/2.Gemini-2.0/rust_solutions/1745-palindrome-partitioning-iv.rs
impl Solution {
    pub fn check_partitioning(s: String) -> bool {
        let n = s.len();
        let s_bytes = s.as_bytes();
        let mut dp = vec![vec![false; n]; n];

        for i in 0..n {
            dp[i][i] = true;
        }

        for i in 0..n - 1 {
            if s_bytes[i] == s_bytes[i + 1] {
                dp[i][i + 1] = true;
            }
        }

        for len in 3..=n {
            for i in 0..=n - len {
                let j = i + len - 1;
                if s_bytes[i] == s_bytes[j] && dp[i + 1][j - 1] {
                    dp[i][j] = true;
                }
            }
        }

        for i in 0..n - 2 {
            for j in i + 1..n - 1 {
                if dp[0][i] && dp[i + 1][j] && dp[j + 1][n - 1] {
                    return true;
                }
            }
        }

        false
    }
}